cmake/f3dOptions.cmake

```diff
@@ -0,0 +1,145 @@
+  list(JOIN _options_lister ",\n  " _options_lister)
+
+  configure_file(
+    "${_f3d_generate_options_INPUT_PUBLIC_HEADER}"
+    "${_f3d_generate_options_DESTINATION}/public/${_f3d_generate_options_OUTPUT_NAME}.h")
+
+  configure_file(
+    "${_f3d_generate_options_INPUT_PRIVATE_HEADER}"
+    "${_f3d_generate_options_DESTINATION}/private/${_f3d_generate_options_OUTPUT_NAME}_tools.h")
+
+endfunction()
+
+#[==[
+@brief internal recursive method use to parse json structure and generate variables
+#]==]
+function(_parse_json_option _top_json)
+  # TODO Add a deprecation mechanism
+  string(JSON _options_length LENGTH ${_top_json})
+  MATH(EXPR _options_length "${_options_length} - 1")
+  foreach(_json_idx RANGE ${_options_length})
+    # Recover the json element name
+    string(JSON _member_name MEMBER ${_top_json} ${_json_idx})
+    # Read the json element
+    string(JSON _cur_json GET ${_top_json} ${_member_name})
+    # Recover its type and default if it is a leaf option
+    string(JSON _option_type ERROR_VARIABLE _type_error GET ${_cur_json} "type")
+    string(JSON _option_default_value ERROR_VARIABLE _default_value_error GET ${_cur_json} "default_value")
+    if(_type_error STREQUAL "NOTFOUND" AND _default_value_error STREQUAL "NOTFOUND")
+       # Leaf option found!
+       set(_option_name "${_option_basename}${_member_name}")
+       # Add an option in the struct based on its type
+       if(_option_type STREQUAL "double_vector")
+         string(APPEND _options_struct "${_option_indent}  std::vector<double> ${_member_name} = {${_option_default_value}};\n")
+         list(APPEND _options_setter "if (name == \"${_option_name}\") opt.${_option_name} = std::get<std::vector<double>>(value)")
+       elseif(_option_type STREQUAL "string")
+         string(APPEND _options_struct "${_option_indent}  std::string ${_member_name} = \"${_option_default_value}\";\n")
+         list(APPEND _options_setter "if (name == \"${_option_name}\") opt.${_option_name} = std::get<std::string>(value)")
+       elseif(_option_type STREQUAL "ratio")
+         string(APPEND _options_struct "${_option_indent} f3d::ratio_t ${_member_name} = f3d::ratio_t(${_option_default_value});\n")
+         list(APPEND _options_setter "if (name == \"${_option_name}\") opt.${_option_name} = std::get<double>(value)")
+       else()
+         string(APPEND _options_struct "${_option_indent}  ${_option_type} ${_member_name} = ${_option_default_value};\n")
+         list(APPEND _options_setter "if (name == \"${_option_name}\") opt.${_option_name} = std::get<${_option_type}>(value)")
+       endif()
+       list(APPEND _options_getter "if (name == \"${_option_name}\") var = opt.${_option_name}")
+       list(APPEND _options_string_setter "if (name == \"${_option_name}\") options_tools::parse(str, opt.${_option_name})")
+       list(APPEND _options_string_getter "if (name == \"${_option_name}\") return options_tools::toString(opt.${_option_name})")
```

Comment:
I'd change one of the names to have either `parse` and `format` or `fromString` and `toString`